File: btypes_primitives/src/main/rust/btypes/src/bset.rs
#![ allow( dead_code, non_snake_case) ]

use crate::bboolean::{IntoBool, BBoolean};
use crate::binteger::{BInt, BInteger};
use crate::bstring::BString;
use crate::bobject::BObject;
use crate::btuple::BTuple;
use crate::orderedhashset::OrderedHashSet as OrdSet;

use std::any::TypeId;
use std::convert::TryInto;
use im::ordset::Iter;
use std::hash::Hash;
use std::collections::LinkedList;
use std::fmt;
use rand::Rng;

pub trait TBSet: BObject {
    type Item: BObject;

    fn as_ord_set(&self) -> OrdSet<Self::Item>;
    fn as_bset(&self) -> &BSet<Self::Item>;
}

pub trait SetLike: BObject {
    fn get_empty() -> Self;
    fn _union(&self, other: &Self) -> Self;
    fn intersect(&self, other: &Self) -> Self;
}

#[derive(Default, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Clone)]
pub struct BSet<T: BObject> {
    set: OrdSet<T>,
}

impl<T: BObject> BObject for BSet<T> {}

impl<T: BObject> TBSet for BSet<T> {
    type Item = T;
    fn as_ord_set(&self) -> OrdSet<Self::Item> { self.set.clone() }
    fn as_bset(&self) -> &BSet<Self::Item> { self }
}

impl<T: BObject> fmt::Display for BSet<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut result = "{".to_owned();
        let mut first = true;
        for e in self.set.iter() {
            if !first { result = result + ", " }
            else { first = false; }
            result = result + &format!("{}", e).to_string();
        }
        result = result + "}";
        return write!(f, "{}", result);
    }
}

impl<'a, T: 'static + BObject> IntoIterator for &'a BSet<T>
where
    T: 'a + Ord,
{
    type Item = &'a T;
    type IntoIter = im::ordset::Iter<'a, T>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}

//TODO: check if replacing cache with mutex works and does not impact permormance too much
unsafe impl<T: BObject> Sync for BSet<T> {}

impl<T: 'static + BObject> BSet<T> {

    pub fn new(mut args: Vec<T>) -> BSet<T> {
        let mut ret: BSet<T> = BSet {
            set: OrdSet::new()
        };
        while !args.is_empty() {
            ret.set.insert(args.remove(0));
        }
        return ret;
    }

    pub fn fromOrdSet(set: OrdSet<T>) -> BSet<T> {
        return BSet { set: set };
    }


    pub fn iter(&self) -> Iter<'_, T> {
        return self.set.iter();
    }

    pub fn size(&self) -> usize {
        return self.set.len();
    }

    pub fn isEmpty(&self) -> bool {
        return self.set.is_empty();
    }

    pub fn contains(&self, o: &T) -> bool {
        return self.set.contains(o);
    }

    pub fn intersect(&self, set: &BSet<T>) -> BSet<T> {
        return BSet{ set: self.set.clone().intersection(set.set.clone()) };
    }

    pub fn difference(&self, set: &BSet<T>) -> BSet<T> {
        return BSet{ set: self.set.clone().relative_complement(set.set.clone()) };
    }

    pub fn _union(&self, set: &BSet<T>) -> BSet<T> {
        return BSet{ set: self.set.clone().union(set.set.clone()) };
    }

    pub fn interval(a: &BInteger, b: &BInteger) -> BSet<BInteger> {
        let mut result: BSet<BInteger> = BSet::new(vec![]);
        for i in a.get_val()..b.get_val()+1 {
            result.set.insert(BInteger::new(i));
        }
        return result;
    }

    pub fn card(&self) -> BInteger {
        return self._size();
    }

    pub fn _size(&self) -> BInteger {
        return BInteger::new(self.set.len().try_into().unwrap());
    }

    pub fn elementOf(&self, object: &T) -> BBoolean {
        return BBoolean::new(self.contains(object));
    }

    pub fn notElementOf(&self, object: &T) -> BBoolean {
        return BBoolean::new(!self.contains(object));
    }

    pub fn subset(&self, set: &BSet<T>) -> BBoolean {
        return BBoolean::new(self.set.is_subset(&set.set));
    }

    pub fn notSubset(&self, set: &BSet<T>) -> BBoolean {
        return BBoolean::new(!self.set.is_subset(&set.set));
    }

    pub fn strictSubset(&self, set: &BSet<T>) -> BBoolean {
        return BBoolean::new(self.set.len() < set.set.len() && self.set.is_subset(&set.set));
    }

    pub fn strictNotSubset(&self, set: &BSet<T>) -> BBoolean {
        return BBoolean::new(self.set.len() == set.set.len() || !self.set.is_subset(&set.set));
    }

    pub fn fin(&self) -> BSet<BSet<T>> {
        return self.pow();
    }

    pub fn pow(&self) -> BSet<BSet<T>> {
        return self._pow(true);
    }

    pub fn fin1(&self) -> BSet<BSet<T>> {
        return self.pow1();
    }

    pub fn pow1(&self) -> BSet<BSet<T>> {
        return self._pow(false);
    }

    fn _pow(&self, with_empty_set: bool) -> BSet<BSet<T>> {
        let mut result: BSet<BSet<T>> = BSet::new(vec![]);
        let start: BSet<T> = BSet::new(vec![]);
        let mut queue: LinkedList<BSet<T>> = LinkedList::new();
        queue.push_back(start.clone());
        if with_empty_set { result = result._union(&BSet::new(vec![start])); }

        while !queue.is_empty() {
            let current_set: BSet<T> = queue.pop_front().unwrap();
            for element in self.set.iter() {
                let next_set: BSet<T> = current_set._union(&BSet::new(vec![element.clone()]));
                let previous_size = result.size();
                result = result._union(&BSet::new(vec![next_set.clone()]));
                if previous_size < result.size() {
                    queue.push_back(next_set);
                }
            }
        }

        return result;
    }

    pub fn nondeterminism(&self) -> T {
        let mut rng = rand::thread_rng();
        return self.set.iter().nth(rng.gen_range(0..self.set.len())).unwrap().clone();
    }

    pub fn equal(&self, other: &BSet<T>) -> BBoolean {
        return BBoolean::new(self.set.eq(&other.set));
    }

    pub fn unequal(&self, other: &BSet<T>) -> BBoolean {
        return BBoolean::new(!self.set.eq(&other.set));
    }

    pub fn subsetOfInteger(&self) -> BBoolean {
        return BBoolean::new(TypeId::of::<BInteger>() == TypeId::of::<T>());
    }

    pub fn strictSubsetOfInteger(&self) -> BBoolean {
        return self.subsetOfInteger();
    }

    pub fn notSubsetOfInteger(&self) -> BBoolean {
        return self.subsetOfInteger().not();
    }

    pub fn equalInteger(&self) -> BBoolean {
        return BBoolean::new(false);
    }

    pub fn unequalInteger(&self) -> BBoolean {
        return BBoolean::new(true);
    }

    pub fn equalNatural(&self) -> BBoolean {
        return BBoolean::new(false);
    }

    pub fn unequalNatural(&self) -> BBoolean {
        return BBoolean::new(true);
    }

    pub fn equalNatural1(&self) -> BBoolean {
        return BBoolean::new(false);
    }

    pub fn unequalNatural1(&self) -> BBoolean {
        return BBoolean::new(true);
    }

    pub fn equalString(&self) -> BBoolean {
        return BBoolean::new(false);
    }

    pub fn unequalString(&self) -> BBoolean {
        return BBoolean::new(true);
    }

    pub fn equalStruct(&self) -> BBoolean {
        return BBoolean::new(false);
    }

    pub fn unequalStruct(&self) -> BBoolean {
        return BBoolean::new(true);
    }

    pub fn subsetOfString(&self) -> BBoolean {
        return BBoolean::new(TypeId::of::<BString>() == TypeId::of::<T>());
    }

    pub fn strictSubsetOfString(&self) -> BBoolean {
        return self.subsetOfString();
    }

    pub fn notSubsetOfString(&self) -> BBoolean {
        return self.subsetOfString().not();
    }

    pub fn notStrictSubsetOfString(&self) -> BBoolean {
        return self.strictSubsetOfString().not();
    }

    pub fn subsetOfStruct(&self) -> BBoolean {
        return self.is_struct();
    }

    pub fn strictsubsetOfStruct(&self) -> BBoolean {
        return self.subsetOfStruct();
    }

    pub fn notsubsetOfStruct(&self) -> BBoolean {
        return self.subsetOfStruct().not();
    }

    pub fn notStrictsubsetOfStruct(&self) -> BBoolean {
        return self.strictsubsetOfStruct().not();
    }

    //rust specific
    pub fn cartesian<T1: 'static + BObject, T2: 'static + BObject>(set_a: &OrdSet<T1>, set_b: &OrdSet<T2>) -> OrdSet<BTuple<T1, T2>> {
        if set_a.is_empty() || set_b.is_empty() {return OrdSet::<BTuple<T1, T2>>::new();}
        return set_a.iter()
            .fold(OrdSet::<BTuple<T1, T2>>::new(),
                  |set, lhs| set_b.iter().cloned()
                                                         .fold(set,
                                                               |tset, rhs| tset.update(BTuple::new(lhs.clone(), rhs))))
    }
}

impl<T: 'static + BInt> BSet<T> {

    pub fn notStrictSubsetOfInteger(&self) -> BBoolean {
        return self.strictSubsetOfInteger().not();
    }

    pub fn subsetOfNatural(&self) -> BBoolean {
        if self.subsetOfInteger().booleanValue() {
            return BBoolean::new(self.set.iter().find(|i| i.get_binteger_value().isNotNatural().booleanValue()).is_none());
        }
        return BBoolean::new(false);
    }

    pub fn strictSubsetOfNatural(&self) -> BBoolean {
        return self.subsetOfNatural();
    }

    pub fn notSubsetOfNatural(&self) -> BBoolean {
        return self.subsetOfNatural().not();
    }

    pub fn notStrictSubsetOfNatural(&self) -> BBoolean {
        return self.strictSubsetOfNatural().not();
    }

    pub fn subsetOfNatural1(&self) -> BBoolean {
        if self.subsetOfInteger().booleanValue() {
            return BBoolean::new(self.set.iter().find(|i| i.get_binteger_value().isNotNatural1().booleanValue()).is_none());
        }
        return BBoolean::new(false);
    }

    pub fn strictSubsetOfNatural1(&self) -> BBoolean {
        return self.subsetOfNatural();
    }

    pub fn notSubsetOfNatural1(&self) -> BBoolean {
        return self.subsetOfNatural1().not();
    }

    pub fn notStrictSubsetOfNatural1(&self) -> BBoolean {
        return self.strictSubsetOfNatural1().not();
    }

    pub fn _min(self: &BSet<T>) -> BInteger {
        return self.set.get_min().unwrap().get_binteger_value();
    }

    pub fn _max(self: &BSet<T>) -> BInteger {
        return self.set.get_max().unwrap().get_binteger_value();
    }
}

impl<T: 'static + BObject> SetLike for BSet<T> {
    fn get_empty() -> Self { BSet::<T>::new(vec![]) }
    fn _union(&self, other: &Self) -> Self { self._union(other) }
    fn intersect(&self, other: &Self) -> Self { self.intersect(other) }
}

impl<T: 'static + SetLike> BSet<T> {
    pub fn unary__union(&self) -> T {
        return self.iter().fold(T::get_empty(), |result, next| result._union(next));
    }

    pub fn unary_intersect(&self) -> T {
        return self.iter().fold(T::get_empty(), |result, next| result.intersect(next));
    }
}

File: btypes_big_integer/src/main/js/btypes/BSet.ts
import {BBoolean} from "./BBoolean.js";
import {BInteger} from "./BInteger.js";
import {BObject} from "./BObject.js";
import {BRelation} from "./BRelation.js";
import {BString} from "./BString.js";
import {BStruct} from "./BStruct.js";
import * as immutable from "../immutable/dist/immutable.es.js";

export class BSet<T extends BObject> implements BObject{

	set: immutable.Set<T>;

	constructor(...args: any[]) {
		if (args.length == 1 && args[0] instanceof immutable.Set) {
			this.set = <immutable.Set<T>> args[0];
		} else {
			this.set = immutable.Set();
			for(let x of args) {
				this.set = this.set.add(x)
			}
		}
	}

	/* Make this class iterable */
	public [Symbol.iterator]() {
		return this.set[Symbol.iterator]();
	}

	toString(): string {
		let sb: string = "{";
		this.set.forEach((element: T) => {
			if(sb.length >1) {
				sb += ", ";
			}
			sb += element.toString();
		});
		sb += "}"
		return sb;
	}

	union(other: BSet<T>): BSet<T> {
		let result = BSet.immutableSetUnion(this.set, other.set);
		return new BSet(result)
	}

    unionForSets<K extends BObject>(): BSet<K> {
        if (this.set.isEmpty()) {
            return new BSet<K>();
        } else {
            return this.set.reduce(
                (a: BSet<K>, e: BSet<K>) => a.union(e),
                new BSet<K>()
            );
        }
    }

    unionForRelations<T1 extends BObject, T2 extends BObject>(): BRelation<T1, T2> {
        if (this.set.isEmpty()) {
            return new BRelation<T1, T2>();
        } else {
            return this.set.reduce(
                (a: BRelation<T1, T2>, e: BRelation<T1, T2>) => a.union(e),
                new BRelation<T1, T2>()
            );
        }
    }

	static immutableSetUnion<R extends BObject>(s1: immutable.Set<R>, s2: immutable.Set<R>): immutable.Set<R>{
		return s1.union(s2);
	}

	static immutableSetDifference<R extends BObject>(s1: immutable.Set<R>, s2: immutable.Set<R>): immutable.Set<R> {
		return s1.subtract(s2);
	}

	static immutableSetIntersection<R extends BObject>(s1: immutable.Set<R>, s2: immutable.Set<R>): immutable.Set<R> {
		return s1.intersect(s2);
	}

	static immutableMapUnion<S extends BObject, T extends BObject>(m1: immutable.Map<S, immutable.Set<T>>,
																   m2: immutable.Map<S, immutable.Set<T>>): immutable.Map<S, immutable.Set<T>> {
		let result: immutable.Map<S, immutable.Set<T>> = m2;
		outer_loop:
		for (let current_element of m1.keys()) {
			for (let result_element of result.keys()) {
				if (current_element.equals(result_element)) {
					let result_set = result.get(result_element);
					let current_set = m1.get(current_element);
					result = result.set(result_element,
						BSet.immutableSetUnion(result_set !=null ? result_set : immutable.Set<T>(),
											   current_set !=null ? current_set: immutable.Set<T>()));
					continue outer_loop;
				}
			}
			result = result.set(current_element, <immutable.Set<T>><unknown> m1.get(current_element))
		}
		return result;
	}

	static immutableMapIntersection<S extends BObject, T extends BObject>(m1: immutable.Map<S, immutable.Set<T>>,
																		  m2: immutable.Map<S, immutable.Set<T>>): immutable.Map<S, immutable.Set<T>> {
		let result = immutable.Map<S, immutable.Set<T>>();
		elem_loop:
		for (let result_element of m2.keys()) {
			for (let current_element of m1.keys()) {
				if (current_element.equals(result_element)) {
					let result_set = m2.get(result_element);
					let current_set = m1.get(current_element);
					result = result.set(result_element,
						BSet.immutableSetIntersection(result_set !=null ? result_set : immutable.Set(),
												      current_set !=null ? current_set: immutable.Set()));
					continue elem_loop;
				}
			}
			result.delete(result_element);
		}
		return result;
	}

	intersect(other: BSet<T>): BSet<T> {
		let result = BSet.immutableSetIntersection(this.set, other.set);
		return new BSet(result)
	}

    intersectForSets<K extends BObject>(): BSet<K> {
        if (this.set.isEmpty()) {
            return new BSet<K>();
        } else {
            return this.set.reduce(
                (a: BSet<K>, e: BSet<K>) => a.intersect(e)
            );
        }
    }

    intersectForRelations<T1 extends BObject, T2 extends BObject>(): BRelation<T1, T2> {
        if (this.set.isEmpty()) {
            return new BRelation<T1, T2>();
        } else {
            return this.set.reduce(
                (a: BRelation<T1, T2>, e: BRelation<T1, T2>) => a.intersect(e)
            );
        }
    }

	difference(other: BSet<T>): BSet<T> {
		let set = BSet.immutableSetDifference(this.set, other.set)
		return new BSet<T>(set);
	}

	card(): BInteger {
		return new BInteger(this.set.size);
	}

	size(): BInteger {
		return new BInteger(this.set.size);
	}

	elementOf(obj: T): BBoolean {
		return new BBoolean(this.set.has(obj));
	}

	notElementOf(obj: T): BBoolean {
		return new BBoolean(!this.set.has(obj));
	}

	subset(other: BSet<T>): BBoolean {
		this_set_loop:
		for (let elem of this.set) {
			for (let other_elem of other.set) {
				if (other_elem.equals(elem)) {
					continue this_set_loop;
				}
			}
			return new BBoolean(false);
		}
		return new BBoolean(true);
	}

	notSubset(other: BSet<T>): BBoolean {
		this_set_loop:
			for (let elem of this.set) {
				for (let other_elem of other.set) {
					if (other_elem.equals(elem)) {
						continue this_set_loop;
					}
				}
				return new BBoolean(true);
			}
		return new BBoolean(false);
	}

	strictSubset(other: BSet<T>): BBoolean {
		return this.size().less(other.size()).and(this.subset(other));
	}

	strictNotSubset(other: BSet<T>): BBoolean {
		return this.size().equal(other.size()).and(this.notSubset(other));
	}

	contains(other: T): boolean {
		return other in this.set;
	}

	containsAll(other: BSet<T>): boolean {
		for (let o of other.set) {
			if (!(o in this.set)) {
				return false
			}
		}
		return true
	}

	isEmpty(): boolean {
		return this.set.size === 0
	}

	equals(other: any): boolean {
        if (!(other instanceof BSet)) {
            return false;
        }
        return this.set.equals(other.set);
	}

	equal(other: any): BBoolean {
	    return new BBoolean(this.equals(other));
	}

	unequal(other: any): BBoolean {
		return new BBoolean(!this.equals(other));
	}

	nondeterminism(): T {
		let values = []
		for (let item of this.set) {
			values.push(item)
		}
		return values[Math.floor(Math.random() * values.length)]
	}

	min(): T {
		return this.set.reduce((a: T, v: T) => {if(a<v){return a} return v});
	}

	max(): T {
		return this.set.reduce((a: T, v: T) => {if(a>v){return a} return v});
	}

	pow(): BSet<BSet<T>> {
		let result = new BSet<BSet<T>>()
		let start = new BSet<T>()
		let queue = [start]
		result = result.union(new BSet<BSet<T>>(start))
		while (queue.length != 0) {
			let currentSet = <BSet<T>> queue.pop()
			for (let element of this.set) {
				let nextSet = currentSet.union(new BSet<T>(element))
				let previousSize = result.size()
				result = result.union(new BSet<BSet<T>>(nextSet))
				if (previousSize < result.size()) {
					queue.push(nextSet)
				}
			}
		}

		return result
	}

	pow1(): BSet<BSet<T>> {
		return this.pow().difference(new BSet(new BSet()))
	}

	// Only finite subsets are supported so fin = pow
	fin(): BSet<BSet<T>> {
		return this.pow();
	}

	fin1(): BSet<BSet<T>> {
		return this.pow1();
	}

	subsetOfInteger(): BBoolean {
		for (let element of this.set) {
			if (element ! instanceof BInteger) {
				return new BBoolean(false);
			}
		}
		return new BBoolean(true);
	}

	strictSubsetOfInteger(): BBoolean {
		return this.subsetOfInteger();
	}

	notSubsetOfInteger(): BBoolean {
		return this.subsetOfInteger().not();
	}

	notStrictSubsetOfInteger(): BBoolean {
		return this.strictSubsetOfInteger().not();
	}

	subsetOfNatural(): BBoolean {
		for (let element of this.set) {
			if (!(element instanceof BInteger && element.isNatural().booleanValue())) {
				return new BBoolean(false);
			}
		}
		return new BBoolean(true)
	}

	strictSubsetOfNatural(): BBoolean {
		return this.subsetOfNatural()
	}

	notSubsetOfNatural(): BBoolean {
		return this.subsetOfNatural().not()
	}

	notStrictSubsetOfNatural(): BBoolean {
		return this.strictSubsetOfNatural().not()
	}

	subsetOfNatural1(): BBoolean {
		for (let element of this.set) {
			if (!(element instanceof BInteger && element.isNatural1().booleanValue())) {
				return new BBoolean(false);
			}
		}
		return new BBoolean(true)
	}

	subsetOfString(): BBoolean {
		for (let element of this.set) {
			if (!(element instanceof BString)) {
				return new BBoolean(false)
			}
		}
		return new BBoolean(true)
	}

	strictSubsetOfString(): BBoolean {
		return this.subsetOfString()
	}

	notSubsetOfString(): BBoolean {
		return this.subsetOfString().not()
	}

	notStrictSubsetOfString(): BBoolean {
		return this.strictSubsetOfString().not()
	}

	subsetOfStruct(): BBoolean {
		for (let element of this.set) {
			if (!(element instanceof BStruct)) {
				return new BBoolean(false);
			}
		}
		return new BBoolean(true);
	}

	strictSubsetOfStruct(): BBoolean {
		return this.subsetOfStruct()
	}

	notSubsetOfStruct(): BBoolean {
		return this.subsetOfStruct().not()
	}

	notStrictSubsetOfStruct(): BBoolean {
		return this.strictSubsetOfStruct().not()
	}

	equalInteger(): BBoolean {
		return new BBoolean(false)
	}

	unequalInteger(): BBoolean {
			return new BBoolean(true)
	}

	equalNatural(): BBoolean {
		return new BBoolean(false)
	}

	unequalNatural(): BBoolean {
		return new BBoolean(true)
	}

	equalNatural1(): BBoolean {
		return new BBoolean(false)
	}

	unequalNatural1(): BBoolean {
		return new BBoolean(true)
	}

	equalString(): BBoolean {
		return new BBoolean(false)
	}

	unequalString(): BBoolean {
		return new BBoolean(true)
	}

	equalStruct(): BBoolean {
		return new BBoolean(false)
	}

	unequalStruct(): BBoolean {
		return new BBoolean(true)
	}

	getSet() : immutable.Set<T> {
		return this.set
	}

	static interval(a: BInteger, b: BInteger): BSet<BInteger> {
        if (b.less(a).booleanValue()) {
            return new BSet();
        }

        let persistentSet =  immutable.Set();
        for(let i = a; i.lessEqual(b).booleanValue(); i = i.plus(new BInteger(1))) {
            persistentSet = persistentSet.add(i);
        }

        return new BSet(persistentSet);
	}


	hashCode(): number {
		return this.set.hashCode()
	}
}
